### PR TITLE
New blog posts

### PR DESCRIPTION
**Pull request by perbu**
Hi.

I had submitted a talk for Dockercon that was rejected. I took the time to write it down as blog post - hopefully you'll find it relevant. Let me know if you think it is appropriate. I can assure you it isn't a tabloid as the last one. :-)

### CODE REVIEW

**Comment by amirmc** [comment]
Thanks for the post!  I'll make some minor edits and push to your branch by tomorrow.  We should be able to get this up by Wednesday or Thursday this week.  :)

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+Let us have a look at something the software industry has been struggling with for a while; running native code in browsers. We’ve always known it is a horrible idea but we’ve still spent hundred of thousands of hours making it a reality. If we only were able to build a properly secure sandbox it’ll be great. Turns out making such a sandbox is really hard and this is why [Java Applets], [ActiveX controllers] and [NaCL] have been abandoned (to be truthful, I’m not certain about why Google decided to discontinue NaCL, they’re a bit opaque on why).
+
+## How can Unikernels allow one to run native code in the browser?
+
+It’s reasonably simple. Modern CPUs provide virtualization capabilities built in. These have good security records and to my knowledge an attack that shows how you can break out of the hardware isolation hasn’t be demonstrated. The closest thing we’ve seen is attacks on the hardware emulators that helt facilitate these virtual machines.
```
Comment (justincormack) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: typo "helt"->"help"

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+We have no plans of implementing this. While it certainly is pretty cool there is little opportunity for basing any commercial success on such an implementation.  The point of this exercise is to show you what the possibilities are. Replacing Linux with a Unikernel is one thing, but the really exciting stuff will happen when we realize that the way we view operating systems is holding us back. The whole idea of having to emulate a whole computer with BIOS, PCI buss and a floppy drive in order to run a virtual machine is pretty silly. Once we can free us from that limitation we can start finding real applications for unikernels. In the mean time they can displace a few Linux VMs.
```
Comment (justincormack) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: buss -> bus

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+excerpt:  When a radically different technology comes along it usually takes time before we figure out how to apply it. 
+shorturl: 
+---
+{% include JB/setup %}
+
+When a radically different technology comes along it usually takes time before we figure out how to apply it. When we had steam engines running factories there was one engine in each factory with a giant driveshaft running through the whole factory. When the electric engine came along people started replacing the giant steam engine with a giant electric motor. It took time before people understood that they could deploy several small motors in different parts of the factory and connect electric cables rather than having a common driveshaft. It takes time to understand the technology and its applicability. 
+
+![Steam engine]({{BASE_PATH}}/images/steam_engine.jpg)
+
+The situration with unikernels is similar. We have this new thing and to some extent we're using it to replace some genereal purpose operating system workload. But we're still very much limited by how we think about operating systems and computers.
```
Comment (mato) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: Typo "genereal" -> "general", "workload" -> "workloads".

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Unikernels are radically different. Naturally the question of the killer app has come up on a number of occasions. As unikernels are quite different from the dominant operating systems of today it isn’t as easy to spot what the killer application for Unikernels will be. Here I’ll try to answer why this is hard.
```
Comment (mato) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "what the killer application for Unikernels will be" -> "what their killer application will be" (no need to repeat).

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md (hunk 1 above)
Comment (mato) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "These have good security records" -> Unclear what this is referring to, hardware virtualization as a technology?

I can understand what you're getting at -- hardware virt in isolation has a good track record, all attacks to date have been on the "monitor" side -- but this needs more expansion for people outside of the unikernel development community to understand.

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+## Defining characteristics of Unikernels
+
+Let’s start with what our charactistics are. Unikernels are fast, secure and small. This makes them a good fit for a number of applications. Unikernel Systems original business plan was to pursue serverless architectures. Booting [MirageOS] in a few milliseconds was the driving characteristic behind the initiative. Galios, through their [HaLVM] unikernel are mainly relying on Unikernels being small and frugal to pack a lot of virtual machines into a small little box in their [Cyberchaff] product. [IncludeOS] is focused on network appliances like firewall, load balancers and web based APIs, relying on security and performance to get adoption.
```
Comment (dsheets) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: s/Galios/Galois/

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+ * Talk HTTP to its origin server. Browsers have HTTP clients built in.
+ * Manipulate the DOM the same way javascript can do.
+ * (Optional - only if we would like this to have persistent state) Manipulate some local storage - contained in a file.
+
+All of these mechanisms can be transported out of the VM through hypercall (like [VMCALL]). The VM will run at native speed and the CPU will trap whenever it needs to push data out of the VM.
+
+The browser needs to provide these three interfaces and the unikernel needs to implement them as well. This is where HTTP without IP comes in. How would paravirtualized HTTP without IP look like? The application builds a HTTP request, puts it in memory and signals the hypervisor that it should deliver a message to the virtual machine monitor (VMM - in this case it will be the browser). Basically a classical struct should do it.  
+
+The browser takes  a careful look at the pointer and the data it is pointing at. This code will need to be properly vetted as a buffer overflow here could potentially allow the unikernel to execute code outside it’s hardware prison.
```
Comment (dsheets) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: s/it's/its/

**Review by mato — COMMENTED** [review]
My personal opinion:

Your "killer app" more or less describes a unikernel-based implementation of NaCl. I don't find that particularly compelling -- the argument for native code vs. JS has mostly resolved itself in favour of the latter by hardware getting faster and JS technology improving (see Webassembly/asm.js). If you're exposing "the DOM" to the (possibly untrusted) unikernel and using the browser for talking HTTP to the (definitely untrusted) network then it's not clear where the security benefits lie.

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+excerpt:  When a radically different technology comes along it usually takes time before we figure out how to apply it. 
+shorturl: 
+---
+{% include JB/setup %}
+
+When a radically different technology comes along it usually takes time before we figure out how to apply it. When we had steam engines running factories there was one engine in each factory with a giant driveshaft running through the whole factory. When the electric engine came along people started replacing the giant steam engine with a giant electric motor. It took time before people understood that they could deploy several small motors in different parts of the factory and connect electric cables rather than having a common driveshaft. It takes time to understand the technology and its applicability. 
+
+![Steam engine]({{BASE_PATH}}/images/steam_engine.jpg)
+
+The situration with unikernels is similar. We have this new thing and to some extent we're using it to replace some general purpose operating system workloads. But we're still very much limited by how we think about operating systems and computers.
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "situration" ↝ "situation"

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Unikernels are radically different. Naturally the question of the killer app has come up on a number of occasions. As unikernels are quite different from the dominant operating systems of today it isn’t as easy to spot what it will be. Here I’ll try to answer why it's hard to spot the killer app.
+
+## Defining characteristics of Unikernels
+
+Let’s start with what our charactistics are. Unikernels are fast, secure and small. This makes them a good fit for a number of applications. Unikernel Systems original business plan was to pursue serverless architectures. Booting [MirageOS] in a few milliseconds was the driving characteristic behind the initiative. Galois, through their [HaLVM] unikernel are mainly relying on Unikernels being small and frugal to pack a lot of virtual machines into a small little box in their [Cyberchaff] product. [IncludeOS] is focused on network appliances like firewall, load balancers and web based APIs, relying on security and performance to get adoption.
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "Systems" ↝ "Systems'"

Comment (yallop) [reply blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "Unikernels" ↝ "unikernels" (downcase, and similarly elsewhere)

Comment (yallop) [reply blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "firewall" ↝ "firewalls"

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+But there is an overlooked characteristic of Unikernels. They are flexible. Unlike Linux and Windows you can easily boot a Unikernel that relies on completely different infrastructure and protocols. Porting something as obscure as SNA to a Unikernel is a lot simpler than porting it to Linux or Windows. Likewise, we could well envision having HTTP support in a unikernel without having an IP stack. This sounds crazy and you’re likely asking yourself why would you ever do such a thing. I’ll get back to why and how, but bear with me for now.
+
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "why would you" ↝ "why you would"

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+Let us have a look at something the software industry has been struggling with for a while; running native code in browsers. We’ve always known it is a horrible idea but we’ve still spent hundred of thousands of hours making it a reality. If we only were able to build a properly secure sandbox it’ll be great. Turns out making such a sandbox is really hard and this is why [Java Applets], [ActiveX controllers] and [NaCL] have been abandoned (to be truthful, I’m not certain about why Google decided to discontinue NaCL, they’re a bit opaque on why).
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "while; " ↝ "while: "

Comment (yallop) [reply blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "it'll be" ↝ "it'd be"

Comment (yallop) [reply blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "NaCL, "  ↝ "NaCL; "

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+## How can Unikernels allow one to run native code in the browser?
+
+It’s reasonably simple. Modern CPUs provide virtualization capabilities with good security built in and we can leverage this hardware to provide isolation between server-supplied code and the client it's running on.
+
+Envision the following. You have a webpage that needs to run native code. It could be some brand new of DRM or some real intensive cryptography that needs the full power your CPU can provide. The webpage downloads a unikernel image and prepares the image. The browser sets up the virtual machine, creates page tables and initializes the unikernel. The CPU doesn’t need to boot this virtual machine in Ring 0, it can run in Ring 3 of the CPU as it will never need to talk to hardware, neither virtualized or physical.
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "brand new of DRM" ↝ ?

Comment (yallop) [reply blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "Ring 0, " ↝ "Ring 0; "

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+## How can Unikernels allow one to run native code in the browser?
+
+It’s reasonably simple. Modern CPUs provide virtualization capabilities with good security built in and we can leverage this hardware to provide isolation between server-supplied code and the client it's running on.
+
+Envision the following. You have a webpage that needs to run native code. It could be some brand new of DRM or some real intensive cryptography that needs the full power your CPU can provide. The webpage downloads a unikernel image and prepares the image. The browser sets up the virtual machine, creates page tables and initializes the unikernel. The CPU doesn’t need to boot this virtual machine in Ring 0, it can run in Ring 3 of the CPU as it will never need to talk to hardware, neither virtualized or physical. 
+
+This VM would not have an IP stack but a small, secure and well defined interface. It can basically do two or three things to communicate with the host:
+
+ * Talk HTTP to its origin server. Browsers have HTTP clients built in.
+ * Manipulate the DOM the same way javascript can do.
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "javascript" ↝ "JavaScript"

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ * (Optional - only if we would like this to have persistent state) Manipulate some local storage - contained in a file.
+
+All of these mechanisms can be transported out of the VM through hypercall (like [VMCALL]). The VM will run at native speed and the CPU will trap whenever it needs to push data out of the VM.
+
+The browser needs to provide these three interfaces and the unikernel needs to implement them as well. This is where HTTP without IP comes in. How would paravirtualized HTTP without IP look like? The application builds a HTTP request, puts it in memory and signals the hypervisor that it should deliver a message to the virtual machine monitor (VMM - in this case it will be the browser). Basically a classical struct should do it.
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "How would ... look like?" ↝ "What would ... look like?"

File: blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+We have no plans of implementing this. While it certainly is pretty cool there is little opportunity for basing any commercial success on such an implementation.  The point of this exercise is to show you what the possibilities are. Replacing Linux with a Unikernel is one thing, but the really exciting stuff will happen when we realize that the way we view operating systems is holding us back. The whole idea of having to emulate a whole computer with BIOS, PCI bus and a floppy drive in order to run a virtual machine is pretty silly. Once we can free us from that limitation we can start finding real applications for unikernels. In the mean time they can displace a few Linux VMs.
```
Comment (yallop) [inline blog/_posts/2017-08-01-the-search-for-the-killer-app-of-unikernels.md]: "mean time" ↝ "meantime"

**Comment by perbu** [comment]
@mato, thanks. That is really the sort of feedback I'm looking for. I cannot claim I know too much about the inner workings of NaCL or the more recent attempt - Webassembly, so I'm stepping somewhat out into unknown territory here.

The point I'm trying to make is really a bit meta (making the case for specialized operating systems where we previously haven't used operating systems) - but if the prime example used is meaningless then I'd like to find a better example to get the point across.

**Comment by perbu** [comment]
@mato, thanks. That is really the sort of feedback I'm looking for. I cannot claim I know too much about the inner workings of NaCL or the more recent attempt - Webassembly, so I'm stepping somewhat out into unknown territory here.

The point I'm trying to make is really a bit meta (making the case for specialized operating systems where we previously haven't used operating systems) - but if the prime example used is meaningless then I'd like to find a better example to get the point across. 

And I certainly don't think running untrusted code in the browser is the killer app for anything. I don't really see the need for it. But _if_ we're gonna do it unikernels might be a good way to do so.

**Comment by perbu** [comment]
Guys, sorry about the spelling mistakes. I could have gotten them fixed in my end - at least knowing that it was going to be published. Next time I'll make sure that either the spelling is improved upon submission or 2) make it clear that I'll deal with the spelling once the content is otherwise acceptable.

**Comment by amirmc** [comment]
@perbu I pushed a few changes to your branch.  Would you prefer to think of a different example to build the post around?  Or do you want to publish this anyway?

**Comment by perbu** [comment]
Thanks.

I this we should push it out. The goal of the post isn't how to implement running of untrusted code in a browser, but more of an example of how unikernels can be applied to a different set of problems. If anybody can come up with a better example we can do a followup - or add some comments to the post.

**Comment by amirmc** [comment]
Ok, I'll merge this tomorrow lunchtime.  I'll make a final few edits for the date and file-names shortly beforehand.

**Pull request merged**